---
title: "Lab_10_Metagenomics_MBI3100A_2022_Assignment_Answers"
output:
  pdf_document: default
date: "2022-11-12"
editor_options:
  markdown:
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r, message=FALSE}
# clear the R environment
rm(list = ls())
```

# Install required libraries

Installing these libraries may take some time.
Try to update all other dependencies when prompted (type "a" and enter).

```{r}

if (!require("BiocManager")) install.packages("BiocManager")

if (!require("phyloseq")) BiocManager::install("phyloseq")

if (!require("microbiomeMarker")) BiocManager::install("microbiomeMarkera")

if (!require("tidyverse")) install.packages("tidyverse")

if (!require("dendextend")) install.packages("dendextend")

```

# Load Libraries

```{r}
library(phyloseq)
library(ggplot2)
library(dplyr)
library(dendextend)
library(microbiomeMarker)
```

# Data import

### Question 1: Import the three files named as 'GP_sp_assignment_otu_table_df.csv', 'GP_sp_assignment_sample_data_df.csv', and 'GP_sp_assignment_tax_table_df.csv' and make a phyloseq object named 'asgmt_physeq'? (2 points)

Please provide the correct file/folder path

```{r}
GP_sp_assignment_otu_table = read.table("./assignment_files/GP_sp_assignment_otu_table_df.csv",
           sep = "\t",
           header =  T,
           row.names = "otus")
my_OTU_table = otu_table(GP_sp_assignment_otu_table, taxa_are_rows = TRUE)
GP_sp_assignment_sample_data = read.table("./assignment_files/GP_sp_assignment_sample_data_df.csv",
           sep = "\t", header =  T,
           row.names = "sampleid")
my_Sample_data = sample_data(GP_sp_assignment_sample_data)
GP_sp_assignment_tax_table = read.table("./assignment_files/GP_sp_assignment_tax_table_df.csv",
           sep = "\t",
           header =  T,
           row.names = "otus")
my_tax_table = tax_table(as.matrix(GP_sp_assignment_tax_table)) 

asgmt_physeq = phyloseq(my_OTU_table, my_Sample_data, my_tax_table)
```

### Print asgmt_physeq

```{r}
print(asgmt_physeq)
```

### Question 2: How many taxa, samples, and sample variables are there in asgmt_physeq? (1 point)

``` md
```

### Question 3: List the categories present in the sample variable "SampleOrigin". (1 point)

```{r}
sample_data(asgmt_physeq)$SampleOrigin %>% as.factor %>% levels()
```

### Question 4: Generate a bar plot for sample vs abundance and facet it for the catagories in SampleOrigin (1 point)

```{r}
 p = plot_bar(asgmt_physeq, x= "Phylum", fill = "Phylum", facet_grid=SampleOrigin~.)
 p + theme(legend.position="none") + geom_bar(stat = "identity")
```

### Question 5: Based on the plot generated in question 4, name all the phylum which have a big difference in abundance between "Feces" and "Freshwater" samples? (1 point)

``` md
```

### Transform the absolute abundance into relative abundance and filter the taxa which have mean relative abundance less than 0.0001

```{r}
# To convert to relative abundance
asgmt_physeq_r  = transform_sample_counts(asgmt_physeq, function(x) x / sum(x) )
# Keep the taxa which have a mean values at least 0.0001
asgmt_physeq_rf = filter_taxa(asgmt_physeq_r, function(x) mean(x) > 0.0001, TRUE)
asgmt_physeq_rf
```

### Question 6: How many taxa are left after the above filtering? (1 point)

``` md
```

# For question 7 to 12, use dataset 'asgmt_physeq'.

### Question 7: Generate a Hierarchical clustering plot using the distance "ward.D2". (2 points)

#### It will be a four step process

#### Step1: Extract OTU table as data frame

```{r}
asgmt_physeq_otu_df = phyloseq::otu_table(asgmt_physeq) %>%  data.frame()
asgmt_physeq_otu_df[1:5, 1:5]
```

#### Step2: Transpose the table (required by vegdist package)

```{r}
# transpose the table (required by vegdist)
asgmt_physeq_otu_df_t = t(asgmt_physeq_otu_df)
asgmt_physeq_otu_df_t[1:5, 1:5]
```

#### Step3: Compute Bray-Curtis dissimilarity

```{r}
#compute Bray-Curtis dissimilarity
bc_dist = vegan::vegdist(asgmt_physeq_otu_df_t, method = "bray")
bc_dist
```

#### Step4: Save as dendrogram

```{r}
#Save as dendrogram
ward = as.dendrogram(hclust(bc_dist, method = "ward.D2"))

#Plot
plot(ward)
```

### Question 8: Plot for alpha diversity using two measures, "Observed" and "Shannon". (1 points)

```{r}
plot_richness(asgmt_physeq, x="SampleOrigin", measures=c("Observed", "Shannon"), color = "SampleOrigin") +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```

### Question 9: Apply wilcox.test to see if the Observed diversity is significantly different for SampleOrigin. (2 points)

#### Step 1: Make a dataframe to combine the outputs of Observed and SampleOrigin.

```{r}
# Make a dataframe to combine the ouputs of Observed, Shannon and SampleOrigin
my_alph_div = data.frame(
  "Observed" = phyloseq::estimate_richness(asgmt_physeq, measures = "Observed"),
  "Shannon" = phyloseq::estimate_richness(asgmt_physeq, measures = "Shannon"),
  "SampleOrigin" = phyloseq::sample_data(asgmt_physeq)$SampleOrigin)
head(my_alph_div)
```

#### Step 2: Check the significance level for wilcox.test

```{r}
#Wilcoxon test for Shannon diversity for categories in SampleOrigin
my_alph_div_wt = wilcox.test(Shannon ~ SampleOrigin, data = my_alph_div, exact = FALSE, conf.int = TRUE)
print(my_alph_div_wt$p.value)
```

##### Note down the p-value? is the difference significant i.e is p-value less than 0.05?

``` md
```

### Question 10: Make a PCoA plot using the "bray" method as distance for beta diversity. (1 point)

```{r}
ordination = ordinate(asgmt_physeq, method="PCoA", distance="bray")
plot_ordination(asgmt_physeq, ordination, color="SampleOrigin") + 
  theme_classic() +
  theme(strip.background = element_blank()) +
  stat_ellipse(linetype = 2)
```

### Question 11: Apply DESeq2 method to identify the differentially abundant taxa based on SampleOrigin column. (1 point)

```{r}
set.seed(2345)
# run_deseq2 command run the program deseq2 to identify DA taxa
# Running this command takes a few seconds
asgmt_physeq_deseq2 = run_deseq2(asgmt_physeq,
                            group = "SampleOrigin",
                            transform = "log10p", # log transformation
                            norm = "rarefy",  # common method for normalization
                            p_adjust = "BH", # adjusted p-value methods
                            )
asgmt_physeq_deseq2
```

### Question 12: Plot the differentially abundant taxa identified by deseq2 method . (1 point)

```{r}
plot_DA =  microbiomeMarker::plot_abundance(asgmt_physeq_deseq2, group = "SampleOrigin")
plot_DA 
```

```{r}
plot_DA_hmap = microbiomeMarker::plot_heatmap(asgmt_physeq_deseq2, group = "SampleOrigin")
plot_DA_hmap
```
