---
title: "Lab_10_Metagenomics_MBI3100A_2022_Tutorial"
output:
  pdf_document: default
date: "2022-11-12"
editor_options:
  markdown:
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

**Acknowledgement:** The commands and datasets used in is tutorial are adapted from the [phyloseq website](https://joey711.github.io/phyloseq/) and [Yan Hui Microbiome tutorial](https://www.yanh.org/2021/01/01/microbiome-r/).

# Install required libraries

Installing these libraries may take some time.
Try to update all other dependencies when prompted (type "a" and enter).

```{r}

if (!require("BiocManager")) install.packages("BiocManager")

if (!require("phyloseq")) BiocManager::install("phyloseq")

if (!require("microbiomeMarker")) BiocManager::install("microbiomeMarkera")

if (!require("tidyverse")) install.packages("tidyverse")

if (!require("dendextend")) install.packages("dendextend")



```

# Load Libraries

```{r}
library(phyloseq)
library(ggplot2)
library(dplyr)
library(dendextend)
library(microbiomeMarker)
```

# List of packages available in phyloseq

Phyloseq come with preloaded datasets.
The datasets can be explored using the following commands.
The column named "Item" contains the list of all the datasets available in phyloseq package

```{r}
phyloseq_datasets = data(package = "phyloseq")
phyloseq_datasets$results
```

```{r}
# To load the GlobalPatterns dataset
data(GlobalPatterns)
```

```{r}
GlobalPatterns

```

# 1. Data exploration

GlobalPatterns is a phyloseq object which contains 1 out_table, 1 sample_data table, 1 taxonomy table and 1 phylogenetic tree.

We can access the different data type of data and table using following commands

```{r}
# otu_table(GlobalPatterns) %>% head()
# sample_data(GlobalPatterns) %>% head()
# tax_table(GlobalPatterns) %>% head()
# phy_tree(GlobalPatterns) %>% head()
```

### For example: How many samples are there in the 'GlobalPatterns' data set?

```{r}
sample_data(GlobalPatterns)
```

``` md
Answer: 26
```

### Access the OTU table from a dataset

##### Since the OTU tables are very big we use the command head to read the top 6 rows of a matrix/dataframe

```{r}
otu_table(GlobalPatterns) %>%  head()
```

##### Another way to see the content of a big table

```{r}
otu_table(GlobalPatterns)[1:5, 1:5]
```

```{r}
sample_data(GlobalPatterns) %>% head()
```

### Access the sample data table and the column content from a dataset

```{r}
# To access the variables in the column 'SampleType'
# The column 'SampleType' is of class factor so get the levels using the command
sample_data(GlobalPatterns)$SampleType %>% levels()
```

#### Read with as.data.frame to see the results in a tabular format

```{r}
sample_data(GlobalPatterns)$SampleType %>% levels() %>% as.data.frame()
```

### How many sample types are available under the SampleType column?

``` md
Answer: 9, "Feces", "Freshwater", "Freshwater (creek)", "Mock", "Ocean"     "Sediment (estuary)", "Skin", "Soil", "Tongue"
```

### Explore the taxonomy table

```{r}
tax_table(GlobalPatterns) %>% head() %>% DT::datatable()
```

```{r}
tax_table(GlobalPatterns) %>% head()
```

# 2. Importing data in R as phyloseq object

We will combine a phyloseq object using otu_table, sample_data and taxonomy file.
Will will read these three file and then combine them to make a phyloseq object to work with them,

```{r}
#data_dir = 
```

## OTU table

```{r}
# To import otus column as rownames, as required by phyloseq
GP_sp_tutorial_otu_table = read.table("./tutorial_files/GP_sp_tutorial_otu_table_df.csv",
           sep = "\t",
           header =  T,
           row.names = "otus") 
GP_sp_tutorial_otu_table[1:5, 1:5]
```

## Sample data

```{r}
# Import with sampleid column as rownames, as required by phyloseq
GP_sp_tutorial_sample_data = read.table("./tutorial_files/GP_sp_tutorial_sample_data_df.csv",
           sep = "\t", header =  T,
           row.names = "sampleid") 
GP_sp_tutorial_sample_data %>% head()
```

## Taxonomy table

```{r}
# To import otus column as rownames, as required by phyloseq
GP_sp_tutorial_tax_table = read.table("./tutorial_files/GP_sp_tutorial_tax_table_df.csv",
           sep = "\t",
           header =  T,
           row.names = "otus") 
GP_sp_tutorial_tax_table %>% head()

```

#### In order to read the OTU table as phyloseq object we need to use the following command

```{r}
my_OTU_table = otu_table(GP_sp_tutorial_otu_table, taxa_are_rows = TRUE)
my_OTU_table %>% head()
```

#### Similarly to read sample data and taxonomy table as phyloseq objects

```{r}
my_Sample_data = sample_data(GP_sp_tutorial_sample_data)
my_Sample_data %>% head()
```

#### Reading taxonomy file as phyloseq object

```{r}
# the taxonomy table is required in matrix format
my_tax_table = tax_table(as.matrix(GP_sp_tutorial_tax_table)) 
my_tax_table %>% head()
```

### Combine to make a phyloseq object.

```{r}
my_physeq = phyloseq(my_OTU_table, my_Sample_data, my_tax_table)
my_physeq
```

'my_physeq' is now a new phyloseq object which contains the data (OTU table, sample data, and taxonomy table) that we just imported.

#### Explore the categories in a sample variable eg SampleType

```{r}
sample_data(my_physeq)$SampleType %>% as.factor %>% levels()
```

```{r}
sample_data(my_physeq)$SampleOrigin %>% as.factor %>% levels()
```

# 3. Data visualization and formatting

### Basic plot

```{r}
 p = plot_bar(my_physeq, fill = "Species")
 p + theme(legend.position="none")
```

```{r}
 p = plot_bar(my_physeq, x= "Phylum", fill = "Phylum", facet_grid=SampleOrigin~.)
 p + theme(legend.position="none") + geom_bar(stat = "identity")
```

## Relative abundance and filtering

Some of the analysis provide better results when we work with relative abundance data (frequency table for OTU numbers).
To avoid the spurious results, we can also filter some taxa which have very low abundance.

```{r}
# To convert to relative abundance
my_physeq_r  = transform_sample_counts(my_physeq, function(x) x / sum(x) )
# Keep the taxa which have a mean values at least 1e-5
my_physeq_rf = filter_taxa(my_physeq_r, function(x) mean(x) > 1e-5, TRUE)
my_physeq_rf
```

Now the number of remaining taxa after filtering low abundance taxa is 511 out of 1413 in the full dataset.

## Plot and compare the relative abundance

Notice the taxa abundant in only one group.

```{r}
phyloseq::plot_bar(my_physeq_rf , fill = "Phylum") +
  geom_bar(aes(color = Phylum, fill = Phylum), stat = "identity", position = "stack") +
  labs(x = "Samples", y = "Relative Abundance\n") +
  facet_wrap(~ SampleOrigin, scales = "free")
```

### An example of how to subset the data based on a condition

Here we will remove all the taxa which have "NA" in the Species column.
i.e.
Species is unknown for these OTUs

```{r}
GP_subset_1 = subset_taxa(GlobalPatterns, Species!="NA")
GP_subset_1
```

### Agglomerate taxa at Class level (required by plot_heatmap option)

```{r}
my_physeq_glom = tax_glom(my_physeq, taxrank="Class")
```

```{r}
phyloseq::plot_heatmap(my_physeq_glom, low = "yellow", high = "red", na.value = "white",taxa.label = "Class") +
  facet_grid(~SampleOrigin, scales = "free_x")
```

# 4. Hierarchical clustering

```{r}
#Extract OTU table as data frame
my_physeq_otu_df = phyloseq::otu_table(my_physeq) %>%  data.frame()
my_physeq_otu_df[1:5, 1:5]
```

```{r}
# transpose the table (required by vegdist)
my_physeq_otu_df_t = t(my_physeq_otu_df)
my_physeq_otu_df_t[1:5, 1:5]
```

```{r}
#compute Bray-Curtis dissimilarity
bc_dist = vegan::vegdist(my_physeq_otu_df_t, method = "bray")
bc_dist
```

```{r}
# View the distance matrix as matrix
as.matrix(bc_dist)[1:5, 1:5]
```

The distance table records the Bray_Curtis distance between all samples.
For example, the distance between M31Plmr and M11Plmr is 0.8201904.
Next this matrix will be used for plotting the dendrogram.
'hclust' is the command for hierarchical clustering on distance matrix

```{r}
#Save as dendrogram
ward = as.dendrogram(hclust(bc_dist, method = "ward.D2"))

#Plot
plot(ward)
```

### A nicer dendrogram plot with color coding

Below is the code to color the samples based on their category in SampleOrigin.

```{r}
#Provide color codes
meta = data.frame(phyloseq::sample_data(my_physeq))
colorCode = c(`Freshwater` = "#F8766D", Human = "#00BFC4" )
labels_colors(ward) = colorCode[meta$SampleOrigin][order.dendrogram(ward)]
#Plot
plot(ward)
```

# 5. Alpha diversity

Now we will plot for alpha diversity using two measures, c("Observed", "Shannon"),

```{r}
plot_richness(my_physeq, x="SampleOrigin", measures=c("Observed", "Shannon"), color = "SampleOrigin") +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```

### Identifying the level of significance for the diversity between Feces and Freshwater

```{r}
# Make a dataframe to combine the ouputs of Observed, Shannon and SampleOrigin
my_alph_div = data.frame(
  "Observed" = phyloseq::estimate_richness(my_physeq, measures = "Observed"),
  "Shannon" = phyloseq::estimate_richness(my_physeq, measures = "Shannon"),
  "SampleOrigin" = phyloseq::sample_data(my_physeq)$SampleOrigin)
head(my_alph_div)
```

### Check the level of significance

```{r}
#Wilcoxon test for Shannon diversity for categories in SampleOrigin
my_alph_div_wt = wilcox.test(Shannon ~ SampleOrigin, data = my_alph_div, exact = FALSE, conf.int = TRUE)
print(my_alph_div_wt$p.value)
```

p-value = 0.6761.
Here the p-value is more than 0.05 If the p-value in the output is less than 0.05, it means the difference is significant.
In this case we can **NOT** accept the alternative hypothesis which means that the diversity is **NOT** significantly different between Freshwater and Human samples.

# 6. Beta diversity

```{r}
ordination = ordinate(my_physeq, method="PCoA", distance="jaccard")
plot_ordination(my_physeq, ordination, color="SampleOrigin") + 
  theme_classic() +
  theme(strip.background = element_blank()) +
  stat_ellipse(linetype = 2)
```

It shows that the between sample diversity is very high between Freshwater and very low between Feces samples

# 7. Differential abundance (DA) analysis using deseq2

```{r}
set.seed(2345)
# run_deseq2 command run the program deseq2 to identify DA taxa
# Running this command takes a few seconds
my_physeq_deseq2 = run_deseq2(my_physeq,
                            group = "SampleOrigin",
                            transform = "log10p", # log transformation
                            norm = "rarefy",  # common method for normalization
                            p_adjust = "BH", # adjusted p-value methods
                            )
my_physeq_deseq2
```

### Plot the differentially abundant taxa identified by deseq2 method

```{r}
marker_table(my_physeq_deseq2) %>% head()
```

```{r}
plot_DA =  microbiomeMarker::plot_abundance(my_physeq_deseq2, group = "SampleOrigin")
plot_DA 
```

```{r}
plot_DA_hmap = microbiomeMarker::plot_heatmap(my_physeq_deseq2, group = "SampleOrigin")
plot_DA_hmap
```

```{r}
#otu_table(my_physeq_deseq2) %>% head()
```
